### ISSUE TITLE
[Feature request] Option to take image captions from filename

### ISSUE THREAD

**Issue by jeanhadrien** [d8ahazard/sd_dreambooth_extension#72]
Maybe this feature requests diverges from the original dreambooth paper, but I've found great results by captioning each image separately instead of using a general instance/class caption. This was done on Kane's repo here :

`https://github.com/kanewallmann/Dreambooth-Stable-Diffusion`

Here is an example of training dataset images, captioned using BLIP :
![image](https://user-images.githubusercontent.com/40248338/200902060-2ca2f0a4-084b-4bb2-ac97-4e261da6c0c5.png)

- a girl and her dog are waiting for a train to pass by on a sunny day with a train going by, by samdoesarts_11065.png
- a girl sitting in a car with a teddy bear in her lap and a dog in the back seat, by samdoesarts_70508.png
- etc.

",by samdoesarts" was added manually at the end of each filename.

With that kind of captioning, you're able to greatly enhance the results of your model.
Here's some examples of a model that, in my opinion, greatly captures samdoearts style.

Portraits :
![azeswaze](https://user-images.githubusercontent.com/40248338/200906330-9814d49e-4b0e-4ab6-8324-e98a30a5dc66.png)

More creative prompts :
![azesqdssdqwaze](https://user-images.githubusercontent.com/40248338/200906240-88a3c067-ee20-4cf4-9bc3-d307cb20c8c1.png)

Is there a possibility that you include this feature in your repo ? 

I would say there are multiple levels at which you can integrate this : 
- Just take existing caption from filename, before the "_" character. This means users are expected to preprocess data by themselves.
- Use the already existing feature on automatic webui that preprocess images and generate txt files with a caption using BLIP. Maybe you would need to add the option to append, prepend, replace tokens, in those generated captions, for instance/class images. For example, being able to append "by samdoesarts", as I did manually in my exemple. If training subjects, replace "a woman"/ "a man" with the name of your subject.

In the json file, you would just need to add a path to the folders and not specify instance/class prompts.

There is also the possibility to generate class images on the fly by using the original caption of each instance image. (for each instance image, generate a class image using the base caption of the instance image = without "by samdoesarts" added or "a man" replaced, etc.). My results show that the final model is much more generalist and will understand concepts much better than if you use a single class caption. 

Able to discuss further on discord if needed 😄 : JH#7326

**Comment by CapPow** [comment]
Glad to see others making this request as well! I've had some success using [TheLastBen's implimentation](https://github.com/TheLastBen/fast-stable-diffusion) to train multiple, sometimes overlapping concepts designated by the filename. This method avoids having to duplicate overlapping images into separate folders, and ensures all appropriate concepts are represented in the prompt every time it is presented for training. For example I can train concepts [A], [B], and [C] using filenames like: "a photo with [A] and [C].jpg" and "a photo with [B] and [C].jpg"

**Comment by d8ahazard** [comment]
YASSS.

I was actually looking at how I could do this last night - scouring the bit that generates the training data and looking at how the main TI trainer uses existing captions as well. 

So, obviously, I'm very interested in this feature. If there's already an example of this in the wild, then it shouldn't be too difficult for me to hog this into the extension as well.

**Comment by sgsdxzy** [comment]
I am working on this, and got reading from directory created by preprocess from webui to work. Can't verify the results as both preview and saving are broken as of now. Need to go to bed, might be able to make a complete pull request tomorrow.

**Comment by jeanhadrien** [comment]
@sgsdxzy @d8ahazard thank you guys for your work and dedication !

**Comment by CapPow** [comment]
@d8ahazard, Thanks for the great work! Looking at [TheLastBen's notebook](https://github.com/TheLastBen/fast-stable-diffusion/blob/main/fast-DreamBooth.ipynb), using the "Captionned_instance_images" option seems to activate a parameter in a custom dreambooth training which is [implemented here](https://github.com/TheLastBen/diffusers/blob/9d08993f106b19e2fe8216ab9fe3d19c57963e57/examples/dreambooth/train_dreambooth.py#L330). Hopefully this helps track down a relevant reference example.

**Comment by CrazyBoyM** [comment]
I make a PR here for training style.  
https://github.com/d8ahazard/sd_dreambooth_extension/pull/92

**Comment by d8ahazard** [comment]
Implemented with #92

**Comment by Jonseed** [comment]
Is it necessary to add the suffix, such as ",by samdoesarts", or can the caption alone be sufficient? Does the suffix help the style be targetable in the prompt?

**Comment by CrazyBoyM** [comment]
> Is it necessary to add the suffix, such as ",by samdoesarts", or can the caption alone be sufficient? Does the suffix help the style be targetable in the prompt?

As many people have test it,  
if you want to train a specific style concept or object, just make sure the instance is a special symbols which did not exist before
such as:
```
1.
instance prompt:  a photo of sks dog  
class prompt: a photo of dog
2.
instance prompt:  a photo of <xxx> cat  
class prompt: a photo of cat
...
the class prompt will be used to generate some prior knowledge images, and the instance prompt is a concept for your special image.
```
The loss will be calculated for both prior knowledge images and your sample images.
```
loss = loss + args.prior_loss_weight * prior_loss
the prior_loss_weight is defaultlt set with 1.  
```
if you want to train a special model for you, no need to get some priori knowledge pictures with the class prompt. 
Just make every image has a single label file for it  
Each pair of label and image will be read, and the whole base model will be thoroughly finetuned by your own dataset.
